### PR TITLE
[BACKLOG-25403] Add websocket to CDF components that use the streamin…

### PR DESCRIPTION
**Pull request by joaormf**
…g CDA datasource (replaces the streaming component refresh)

### CODE REVIEW

**Comment by joaormf** [comment]
@plagoa @tgf @marcoslarsen @JoaoRPF @webdetails/millenniumfalcon please review.

Don't merge this one without pentaho/pdi-dataservice-plugin#165 and pentaho/pdi-dataservice-server-plugin#327 being merged.

This should also not be merged without merging the https://github.com/webdetails/cda/pull/281 and https://github.com/webdetails/cde/pull/901

File: core-js/src/main/javascript/cdf/dashboard/RefreshEngine.js [hunk 1]
```diff
@@ -289,15 +289,6 @@ define([
         //tries to fetch the component refresh rate from the data source definition
         var dataSourceId = component.chartDefinition && component.chartDefinition.dataSource ? dataSourceId = component.chartDefinition.dataSource :
```
Comment (joaormf) [inline core-js/src/main/javascript/cdf/dashboard/RefreshEngine.js]: This variable is not needed... I should remove it in a new commit into this PR

File: core-js/src/main/javascript/cdf/queries/CdaQuery.js [hunk 2]
```diff
@@ -348,7 +349,47 @@ define([
...
+    doQuery: function(successCallback, errorCallback) {
+      if(!Utils.isFunction(this.getOption('successCallback'))) {
+        throw 'QueryNotInitialized';
+      }
+
+      if(!!this.getOption('pushEnabled')) {
+        var myself = this;
+        var data = undefined;
+
+        var exampleSocket = new WebSocket(this.getOption('websocketUrl'), 'JSON-CDA-query');
```
[the remark this answers is not in the document]
Comment (joaormf) [reply core-js/src/main/javascript/cdf/queries/CdaQuery.js]: I've change it already!!!

File: core-js/src/main/javascript/cdf/queries/CdaQuery.js [hunk 3]
```diff
@@ -348,7 +349,46 @@ define([
...
+     *
+     * @param {function} [successCallback] Success callback.
+     * @param {function} [errorCallback]   Error callback.
+     */
+    doQuery: function(successCallback, errorCallback) {
+      if(!Utils.isFunction(this.getOption('successCallback'))) {
+        throw 'QueryNotInitialized';
+      }
+
+      if(!!this.getOption('pushEnabled')) {
```
[the remark this answers is not in the document]
Comment (joaormf) [reply core-js/src/main/javascript/cdf/queries/CdaQuery.js]: Fixed in next commit

File: core-js/src/main/javascript/cdf/components/UnmanagedComponent.js [hunk 4]
```diff
@@ -293,12 +293,13 @@ define([
      *               or {@link cdf.components.UnmanagedComponent#triggerAjax|triggerAjax}),
      *               or explicitly call this method immediately before yielding control back to CDF.</p>
      *
+     * @param {boolean} isDataPush If the lifecycle event is being triggered by a push data event.
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/UnmanagedComponent.js]: `postExec` should not receive the `isDataPush` argument.

File: core-js/src/main/javascript/cdf/components/UnmanagedComponent.js [hunk 5]
```diff
@@ -363,13 +364,18 @@ define([
      *              <p>A component that actually begins execution, by returning `true` from this method, should later complete the
      *              lifecycle by calling either {@link cdf.components.UnmanagedComponent#endExec|endExec} or
      *              {@link cdf.components.UnmanagedComponent#failExec|failExec}.</p>
-     *
+     * @param {boolean} isDataPush If the lifecycle event is being triggered by a push data event.
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/UnmanagedComponent.js]: Probably change `lifecycle event` to simply `lifecycle`.

File: core-js/src/main/javascript/cdf/components/UnmanagedComponent.js [hunk 6]
```diff
@@ -707,7 +720,13 @@ define([
         counter  = this.callCounter();
       }
 
-      function fetchDataSuccessHandler(data) {
+      function fetchDataSuccessHandler(data, isDataPush) {
+        if(isDataPush) {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/UnmanagedComponent.js]: Similar code needs to be added to `getErrorHandler`.

File: core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js [hunk 7]
```diff
@@ -114,25 +115,55 @@
...
+          myself.websocket.send(JSON.stringify(queryDefinition));
+        };
+
+        this.websocket.onmessage = function (event) {
+          var isDataPush = !isFirstTime;
+          isFirstTime = false;
+
+          myself.getSuccessHandler(function (data) {
+            return successHandler(data, isDataPush);
+          })(JSON.parse(event.data));
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js]: `, isDataPush` should be moved to after `JSON.parse(event.data)`:

```js
myself.getSuccessHandler(successHandler)(JSON.parse(event.data), isDataPush));
```

Which then prompts us to move `myself.getSuccessHandler(successHandler)` tranformation directly to the variable `successHandler` above.

File: core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js [hunk 8]
```diff
@@ -114,25 +115,55 @@
...
+        };
+
+        this.websocket.onerror = function (event) {
+          var isDataPush = !isFirstTime;
+          isFirstTime = false;
+
+          myself.getErrorHandler(function (resp, txtStatus, error) {
+            errorHandler(resp, txtStatus, error, isDataPush);
+          })(event.data);
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js]: Idem.

File: core-js/src/main/javascript/cdf/queries/CdaQuery.js [hunk 9]
```diff
@@ -348,6 +355,73 @@ define([
...
+
+        this.websocket.onopen = function (event) {
+          myself.websocket.send(JSON.stringify(queryDefinition));
+        };
+
+        this.websocket.onmessage = function (event) {
+          var isDataPush = !isFirstTime;
+          isFirstTime = false;
+
+          myself.getSuccessHandler(function (data) {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/queries/CdaQuery.js]: See comments above.

File: core-js/src/main/javascript/cdf/queries/CdaQuery.js [hunk 10]
```diff
@@ -348,6 +355,73 @@ define([
...
+            return successHandler(data, isDataPush);
+          })(JSON.parse(event.data));
+        };
+
+        this.websocket.onerror = function (event) {
+          var isDataPush = !isFirstTime;
+          isFirstTime = false;
+
+          myself.getErrorHandler(function (resp, txtStatus, error) {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/queries/CdaQuery.js]: Idem.

File: pentaho-js/src/main/javascript/cdf-legacy/cdf-base.js [hunk 11]
```diff
@@ -52,6 +52,10 @@ wd.cdf.helper = {
         fullPath += "/" + clean(options.action);
     }
     return fullPath;
+  },
+
+  httpUrlToWebSockeUrl: function( url ) {
```
Comment (dcleao) [inline pentaho-js/src/main/javascript/cdf-legacy/cdf-base.js]: Typo?

File: pentaho-js/src/main/javascript/cdf-legacy/components/table.js [hunk 12]
```diff
@@ -249,7 +252,13 @@ var TableComponent = UnmanagedComponent.extend({
         var handler = this.getSuccessHandler(success);
 
         this.queryState.setAjaxOptions({async:true});
-        this.queryState.fetchData(this.parameters == undefined ? [] : this.parameters, handler, this.failureCallback);
+        var params = this.parameters == undefined ? [] : this.parameters;
```
Comment (dcleao) [inline pentaho-js/src/main/javascript/cdf-legacy/components/table.js]: Shouldn't we be also modifying `core.js`'s `getSuccessHandler` and `getErrorHandler`?

File: core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js (hunk 7 above)
Comment (dcleao) [reply core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js]: Ignore.

File: core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js (hunk 8 above)
Comment (dcleao) [reply core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js]: Ignore.

File: core-js/src/main/javascript/cdf/queries/CdaQuery.js (hunk 9 above)
Comment (dcleao) [reply core-js/src/main/javascript/cdf/queries/CdaQuery.js]: Ignore.

File: core-js/src/main/javascript/cdf/queries/CdaQuery.js (hunk 10 above)
Comment (dcleao) [reply core-js/src/main/javascript/cdf/queries/CdaQuery.js]: Ignore.

**Pull request merged**